Replace rethrowing map with generic typed throws

DougGregor opened the pull request:
Adopt typed throws for the `map` operation to propagate thrown error
types through the `map` API. The new `map` has a signature like this:

```swift
  public func map<T, E>(
    _ transform: (Element) throws(E) -> T
  ) throws(E) -> [T]
```

This is done in a manner that is backward compatible for almost all cases:

* The new typed-throws entrypoint is `@_alwaysEmitIntoClient` so it can back-deploy all the way back.
* The old `rethrows` entrypoint is left in place, using `@usableFromInline` with `internal` so it is part of the ABI but not the public interface, `@_silgen_name` so it has the right name in the binary, and a new name `__rethrows_map` so it's otherwise out of the way. The old entrypoint is implemented in terms of the new one.

Note that the implementation details for the existential collection "box" classes rely on method overriding of `_map` operations, and the types are frozen, so  we don't get to change their signatures. However, these are only implementations, not API: the actual API `map` functions can be upgraded to typed throws.

Note that this code makes use of a known hole in `rethrows` checking to allow calling between `rethrows` and typed throws. We'll need to do something about this for source-compatibility reasons, but I'll follow up with that separately.

DougGregor committed:
Replace rethrowing `map` with generic typed throws

Adopt typed throws for the `map` operation to propagate thrown error
types through the `map` API. This is done in a manner that is backward
compatible for almost all cases:

* The new typed-throws entrypoint is `@_alwaysEmitIntoClient` so it
can back-deploy all the way back.
* The old `rethrows` entrypoint is left in place, using
`@usableFromInline` with `internal` so it is part of the ABI but not
the public interface, and `@_disfavoredOverload` so the type checker
avoids it while building the standard library itself. The old
entrypoint is implemented in terms of the new one.

Note that the implementation details for the existential collection
"box" classes rely on method overriding of `_map` operations, and the
types are frozen, so  we don't get to change their signatures. However,
these are only implementations, not API: the actual API `map`
functions can be upgraded to typed throws.

Note that this code makes use of a known hole in `rethrows` checking
to allow calling between `rethrows` and typed throws. We'll need to do
something about this for source-compatibility reasons, but I'll follow
up with that separately.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
Note that we expect some tests to fail here, because I haven't updated the expected output in a number of places.

DougGregor commented:
(Only the last two commits are actually new)

stephentyrone commented:
Generally we provide these legacy ABI entrypoints via `@available(*, unavailable)` rather than `internal @usableFromInline @_disfavoredOverload`. Both ways seem fine to me, but is there any reason for the difference here?

DougGregor commented:
> Generally we provide these legacy ABI entrypoints via `@available(*, unavailable)` (sometimes with a source name like `_legacy_se04xx_map` and `@_silgen_name` to the right symbol) rather than `internal @usableFromInline @_disfavoredOverload`. Both ways seem fine to me, but is there any reason for the difference here?

I prefer `@usableFromInline internal` to `@available(*, unavailable)` because it's more hidden from the language model.

The `@_disfavoredOverload` hack is a new experiment, partly out of laziness of not wanting to look up the SIL names for the existing symbols. I think I've decided that I don't really like it much, and will go back to double-unscored versions for the ABI-only entrypoints.

DougGregor committed:
Improve ABI-only rethrowing `map` shims based on code review

Replace the hackish use of `@_disfavoredOverload` with the more principled
use of `@_silgen_name` for the entrypoint we are maintaining, then rename
these functions in source to `__rethrows_map` to indicate what they're for.

While here, make them `throws` instead of `rethrows`. The ABI is the
same, and `throws` allows us do avoid to do/catch tricks with rethrows
functions.

DougGregor committed:
[Existential collections] Switch internal _map rethrows methods to throws

These class methods are internal, but because they are overridden and are
part of a `@usableFromInline`, `@_fixed_layout` class, we they can't be
moved over to typed throws without breaking ABI. However, they are
only ever called from typed-throws functions, which already need a
do...catch dance to downcast the error itself. Make them `throws`
instead, which is ABI-compatible, but eliminates the need for
do...catch hackery in the function itself.

DougGregor commented:
@swift-ci please smoke test

glessard commented on stdlib/public/core/ExistentialCollection.swift:
```diff
@@ -1303,10 +1303,26 @@ extension AnySequence {
+    _ transform: (Element) throws(E) -> T
+  ) throws(E) -> [T] {
+    do {
+      return try _box._map(transform)
```
Would there be an advantage to adding a typed-error version of `ExistentialCollection._map()` ?

DougGregor replied:
I don't think we can, because it's a fixed-layout type

glessard replied:
According to `UnderscoredAttributes.md` (if it's correct), new virtual methods can be added. It would avoid the catch-and-downcast dance, and if that happened to generate better code it might be worth it.

DougGregor replied:
These methods are `@inlineable`, which implies `@usableFromInline`, so they can't go away. Additionally, their callers need to be `@_alwaysEmitIntoClient`, so we couldn't call into different or new entry points for a typed map without having a dynamic `if #available` check, which seems worse than this forced cast.

I think what we have is the best we're going to get.

glessard commented on stdlib/public/core/Misc.swift:
This is no longer used in the current version of the PR

DougGregor replied:
Ooohhh, you're right! Thank you

DougGregor commented:
How fun. It's very crashy on Intel, yet fine on Apple Silicon. Time for some low-level debugging.

tshortli commented:
> Generally we provide these legacy ABI entrypoints via `@available(*, unavailable)` (sometimes with a source name like `_legacy_se04xx_map` and `@_silgen_name` to the right symbol) rather than `internal @usableFromInline @_disfavoredOverload`. Both ways seem fine to me, but is there any reason for the difference here?

We will need to be careful about this pattern going forward because one of the features I've been working on that was recently enabled on `main` compiles `@available(*, unavailable)` code either down to stubs that trap in the case of a library evolution enabled module, or completely omits the unavailable code entirely for non-resilient modules. This is to save code size in libraries with a lot of unavailable code. Semantically, `@available(*, unavailable)` means to the compiler that "this declaration has never been reachable" so it's not really appropriate for stubs needed for ABI compatibility. I personally think the `@usableFromInline internal` pattern for compatibility stubs is a better one because it doesn't allow clients to even name the stub declaration.

stephentyrone commented:
That’s a massively ABI breaking change that will need to go through evolution, even if it is desirable.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
x86 miscompile is addressed by https://github.com/apple/swift/pull/69819

DougGregor committed:
Update tests for map switching to typed throws

DougGregor committed:
Avoid use of typed throws in SIL opaque values test

SIL opaque values do not yet support typed throws, so avoid the API that
uses them.

DougGregor commented:
A couple more cleanups peeled off this and put into https://github.com/apple/swift/pull/69824

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
macOS toolchain: https://ci.swift.org/job/swift-PR-toolchain-macos/945/
Linux toolchain: https://ci.swift.org/job/swift-PR-toolchain-Linux/643/
Windows toolchain: https://ci-external.swift.org/job/swift-PR-build-toolchain-windows/1020/

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please build toolchain

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
### Performance (x86_64): -O

**Regression**                          | **OLD**  | **NEW**  | **DELTA** | **RATIO**
:---                                    | ---:     | ---:     | ---:      | ---:     
MapReduce                               | 92.105   | 11138.0  | +11992.6% | **0.01x**
MapReduceClass2                         | 13.042   | 1302.0   | +9882.4%  | **0.01x**
MapReduceAnyCollection                  | 148.429  | 12052.0  | +8019.7%  | **0.01x**
RangeAssignment                         | 152.333  | 5489.0   | +3503.3%  | **0.03x**
MapReduceSequence                       | 216.444  | 5436.0   | +2411.5%  | **0.04x**
MapReduceClassShort2                    | 95.192   | 1450.0   | +1423.2%  | **0.07x**
MapReduceAnyCollectionShort             | 1000.0   | 14690.0  | +1369.0%  | **0.07x**
MapReduceShort                          | 925.882  | 13200.0  | +1325.7%  | **0.07x**
MapReduceNSDecimalNumber                | 211.182  | 1357.0   | +542.6%   | **0.16x**
MapReduceNSDecimalNumberShort           | 295.143  | 1539.0   | +421.4%   | **0.19x**
MapReduceString                         | 44.455   | 145.25   | +226.7%   | **0.31x**
DictOfArraysToArrayOfDicts              | 394.2    | 1236.0   | +213.5%   | **0.32x**
MapReduceShortString                    | 10.886   | 25.741   | +136.4%   | **0.42x**
WordCountHistogramASCII                 | 2498.113 | 4890.0   | +95.7%    | **0.51x**
WordCountHistogramUTF16                 | 2771.831 | 5188.889 | +87.2%    | **0.53x**
CStringShortAscii                       | 919.091  | 1678.333 | +82.6%    | **0.55x**
CxxStringConversion.cxx.to.swift        | 131.0    | 169.0    | +29.0%    | **0.78x (?)**
EqualSubstringString                    | 23.075   | 29.73    | +28.8%    | **0.78x (?)**
LessSubstringSubstring                  | 23.184   | 29.617   | +27.7%    | **0.78x (?)**
LessSubstringSubstringGenericComparable | 23.303   | 29.708   | +27.5%    | **0.78x (?)**
EqualSubstringSubstring                 | 23.531   | 29.941   | +27.2%    | **0.79x (?)**
EqualSubstringSubstringGenericEquatable | 23.5     | 29.719   | +26.5%    | **0.79x (?)**
EqualStringSubstring                    | 23.609   | 29.714   | +25.9%    | **0.79x (?)**
UTF8Decode_InitFromBytes                | 143.182  | 180.182  | +25.8%    | **0.79x (?)**
CSVParsing.Scalar                       | 355.0    | 446.333  | +25.7%    | **0.80x (?)**
ArrayAppendUTF16Substring               | 11382.0  | 14227.2  | +25.0%    | **0.80x (?)**
ArrayAppendLatin1Substring              | 11634.0  | 14508.0  | +24.7%    | **0.80x (?)**
UTF8Decode_InitFromData                 | 139.154  | 172.417  | +23.9%    | **0.81x (?)**
RomanNumbers2                           | 459.25   | 564.0    | +22.8%    | **0.81x (?)**
ArrayAppendAsciiSubstring               | 11358.0  | 13869.0  | +22.1%    | **0.82x (?)**
SortIntPyramid                          | 492.308  | 585.385  | +18.9%    | **0.84x (?)**
PrefixWhileAnySequence                  | 190.4    | 222.5    | +16.9%    | **0.86x (?)**
ObjectiveCBridgeStubDateAccess          | 130.733  | 152.5    | +16.6%    | **0.86x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed     | 197.9    | 229.5    | +16.0%    | **0.86x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed     | 202.636  | 234.889  | +15.9%    | **0.86x (?)**
Set.isDisjoint.Box.Empty                | 48.656   | 55.828   | +14.7%    | **0.87x (?)**
Set.isDisjoint.Seq.Box.Empty            | 78.4     | 89.286   | +13.9%    | **0.88x (?)**
PrefixWhileSequence                     | 189.182  | 213.667  | +12.9%    | **0.89x (?)**
Set.isSubset.Int.Empty                  | 45.63    | 51.417   | +12.7%    | **0.89x (?)**
DropLastAnySequence                     | 302.833  | 340.667  | +12.5%    | **0.89x (?)**
DropLastSequence                        | 303.833  | 341.2    | +12.3%    | **0.89x (?)**
DropLastSequenceLazy                    | 305.143  | 340.0    | +11.4%    | **0.90x (?)**
SortAdjacentIntPyramids                 | 827.273  | 920.0    | +11.2%    | **0.90x (?)**
ObjectiveCBridgeStringHash              | 62.188   | 69.133   | +11.2%    | **0.90x (?)**
Combos                                  | 235.571  | 260.167  | +10.4%    | **0.91x (?)**
Set.isDisjoint.Int.Empty                | 44.653   | 49.0     | +9.7%     | **0.91x (?)**
StringDistance.characters.ascii         | 2577.0   | 2814.0   | +9.2%     | **0.92x (?)**
NSStringConversion.InlineBuffer.UTF8    | 613.5    | 660.5    | +7.7%     | **0.93x (?)**
&nbsp; | | | | 
**Improvement**                         | **OLD**  | **NEW**  | **DELTA** | **RATIO**
Dictionary4                             | 188.889  | 153.7    | -18.6%    | **1.23x (?)**
Array.removeAll.keepingCapacity.Object  | 6.661    | 5.448    | -18.2%    | **1.22x (?)**
Dictionary4OfObjects                    | 219.25   | 183.714  | -16.2%    | **1.19x (?)**
NormalizedIterator_fastPrenormal        | 541.304  | 485.49   | -10.3%    | **1.11x (?)**
StringComparison_ascii                  | 198.0    | 179.077  | -9.6%     | **1.11x (?)**
NormalizedIterator_emoji                | 363.538  | 330.069  | -9.2%     | **1.10x (?)**
ObjectiveCBridgeStubFromNSStringRef     | 107.222  | 99.467   | -7.2%     | **1.08x (?)**

### Code size: -O

**Regression**               | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                         | ---:   | ---:   | ---:   | ---:   
Combos.o                     | 4822   | 5886   | +22.1% | **0.82x**
FlattenList.o                | 3556   | 4293   | +20.7% | **0.83x**
RangeOverlaps.o              | 5267   | 6349   | +20.5% | **0.83x**
RangeAssignment.o            | 3318   | 3912   | +17.9% | **0.85x**
RomanNumbers.o               | 4284   | 4776   | +11.5% | **0.90x**
LazyFilter.o                 | 7055   | 7787   | +10.4% | **0.91x**
ArrayRemoveAll.o             | 7899   | 8586   | +8.7%  | **0.92x**
DictionaryGroup.o            | 12125  | 13139  | +8.4%  | **0.92x**
ReversedCollections.o        | 8559   | 9242   | +8.0%  | **0.93x**
AngryPhonebook.o             | 8831   | 9474   | +7.3%  | **0.93x**
CString.o                    | 5632   | 6040   | +7.2%  | **0.93x**
StringComparison.o           | 34803  | 37033  | +6.4%  | **0.94x**
BucketSort.o                 | 8227   | 8731   | +6.1%  | **0.94x**
QueueTest.o                  | 11613  | 12316  | +6.1%  | **0.94x**
StringEdits.o                | 9012   | 9487   | +5.3%  | **0.95x**
RemoveWhere.o                | 13911  | 14630  | +5.2%  | **0.95x**
ArrayAppend.o                | 25855  | 26945  | +4.2%  | **0.96x**
WordCount.o                  | 36530  | 37463  | +2.6%  | **0.98x**
FloatingPointConversion.o    | 31340  | 31848  | +1.6%  | **0.98x**
Breadcrumbs.o                | 43827  | 44427  | +1.4%  | **0.99x**
&nbsp; | | | | 
**Improvement**              | **OLD** | **NEW** | **DELTA** | **RATIO**
IntegerParsing.o             | 66995  | 65073  | -2.9%  | **1.03x**
MapReduce.o                  | 23659  | 23156  | -2.1%  | **1.02x**
DictOfArraysToArrayOfDicts.o | 21763  | 21341  | -1.9%  | **1.02x**
DriverUtils.o                | 148751 | 147079 | -1.1%  | **1.01x**

### Performance (x86_64): -Osize

**Regression**                                         | **OLD**  | **NEW**  | **DELTA** | **RATIO**
:---                                                   | ---:     | ---:     | ---:      | ---:     
MapReduce                                              | 109.533  | 11222.0  | +10145.2% | **0.01x**
MapReduceAnyCollection                                 | 144.571  | 12045.0  | +8231.5%  | **0.01x**
RangeAssignment                                        | 157.286  | 5544.0   | +3424.8%  | **0.03x**
MapReduceSequence                                      | 278.286  | 5486.0   | +1871.3%  | **0.05x**
MapReduceShort                                         | 995.556  | 13930.0  | +1299.2%  | **0.07x**
MapReduceAnyCollectionShort                            | 1130.0   | 14420.0  | +1176.1%  | **0.08x**
MapReduceClass2                                        | 138.471  | 1278.0   | +822.9%   | **0.11x**
MapReduceClassShort2                                   | 220.4    | 1485.0   | +573.8%   | **0.15x**
MapReduceNSDecimalNumber                               | 211.9    | 1361.0   | +542.3%   | **0.16x**
MapReduceNSDecimalNumberShort                          | 302.0    | 1553.0   | +414.2%   | **0.19x**
MapReduceString                                        | 44.159   | 154.083  | +248.9%   | **0.29x**
DictOfArraysToArrayOfDicts                             | 410.0    | 1290.0   | +214.6%   | **0.32x**
MapReduceShortString                                   | 11.0     | 29.034   | +163.9%   | **0.38x**
WordCountHistogramASCII                                | 2726.0   | 5162.069 | +89.4%    | **0.53x (?)**
CStringShortAscii                                      | 917.143  | 1680.0   | +83.2%    | **0.55x**
WordCountHistogramUTF16                                | 3031.25  | 5416.667 | +78.7%    | **0.56x**
UTF8Decode_InitFromCustom_noncontiguous_ascii          | 589.0    | 837.0    | +42.1%    | **0.70x**
UTF8Decode_InitFromCustom_noncontiguous_ascii_as_ascii | 677.667  | 948.0    | +39.9%    | **0.71x (?)**
RomanNumbers2                                          | 443.2    | 568.667  | +28.3%    | **0.78x**
LessSubstringSubstring                                 | 23.188   | 29.72    | +28.2%    | **0.78x (?)**
UTF8Decode_InitFromCustom_noncontiguous                | 286.25   | 366.833  | +28.2%    | **0.78x (?)**
EqualSubstringString                                   | 23.485   | 30.042   | +27.9%    | **0.78x (?)**
EqualSubstringSubstringGenericEquatable                | 22.864   | 29.182   | +27.6%    | **0.78x (?)**
EqualSubstringSubstring                                | 22.86    | 29.16    | +27.6%    | **0.78x (?)**
EqualStringSubstring                                   | 22.96    | 29.176   | +27.1%    | **0.79x (?)**
LessSubstringSubstringGenericComparable                | 23.088   | 29.281   | +26.8%    | **0.79x (?)**
CSVParsing.Scalar                                      | 354.25   | 448.0    | +26.5%    | **0.79x (?)**
UTF8Decode_InitFromData                                | 140.0    | 176.3    | +25.9%    | **0.79x (?)**
UTF8Decode_InitFromBytes                               | 142.0    | 177.7    | +25.1%    | **0.80x (?)**
StringComparison_longSharedPrefix                      | 203.5    | 241.0    | +18.4%    | **0.84x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed                    | 197.5    | 229.8    | +16.4%    | **0.86x (?)**
PrefixWhileSequence                                    | 182.5    | 212.111  | +16.2%    | **0.86x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed                    | 201.3    | 233.0    | +15.7%    | **0.86x (?)**
Set.subtracting.Empty.Box                              | 12.635   | 14.486   | +14.6%    | **0.87x (?)**
PrefixWhileAnySequence                                 | 181.727  | 208.111  | +14.5%    | **0.87x (?)**
SortIntPyramid                                         | 522.917  | 597.273  | +14.2%    | **0.88x (?)**
ArrayInClass                                           | 174.925  | 196.25   | +12.2%    | **0.89x (?)**
SortAdjacentIntPyramids                                | 770.0    | 858.75   | +11.5%    | **0.90x (?)**
DistinctClassFieldAccesses                             | 39.344   | 43.63    | +10.9%    | **0.90x (?)**
Combos                                                 | 241.429  | 267.5    | +10.8%    | **0.90x (?)**
ObjectiveCBridgeStringHash                             | 62.222   | 68.87    | +10.7%    | **0.90x (?)**
StringDistance.characters.ascii                        | 2578.0   | 2816.0   | +9.2%     | **0.92x (?)**
Array2D                                                | 4156.444 | 4528.0   | +8.9%     | **0.92x (?)**
RemoveWhereFilterInts                                  | 23.625   | 25.69    | +8.7%     | **0.92x (?)**
NSStringConversion.InlineBuffer.UTF8                   | 655.0    | 710.0    | +8.4%     | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                                        | **OLD**  | **NEW**  | **DELTA** | **RATIO**
MapReduceLazyCollectionShort                           | 67.524   | 31.24    | -53.7%    | **2.16x**
MapReduceLazySequence                                  | 65.929   | 44.091   | -33.1%    | **1.50x**
Array.removeAll.keepingCapacity.Object                 | 6.962    | 5.63     | -19.1%    | **1.24x (?)**
ObjectiveCBridgeStubDateAccess                         | 152.5    | 130.727  | -14.3%    | **1.17x (?)**
StringComparison_ascii                                 | 207.2    | 183.462  | -11.5%    | **1.13x (?)**
NormalizedIterator_nonBMPSlowestPrenormal              | 469.615  | 418.889  | -10.8%    | **1.12x (?)**
NormalizedIterator_emoji                               | 358.37   | 330.759  | -7.7%     | **1.08x (?)**
NormalizedIterator_latin1                              | 185.2    | 172.5    | -6.9%     | **1.07x (?)**

### Code size: -Osize

**Regression**            | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                      | ---:   | ---:   | ---:   | ---:   
Combos.o                  | 4834   | 5943   | +22.9% | **0.81x**
FlattenList.o             | 3308   | 4065   | +22.9% | **0.81x**
RangeAssignment.o         | 2881   | 3536   | +22.7% | **0.81x**
RangeOverlaps.o           | 4779   | 5828   | +22.0% | **0.82x**
RomanNumbers.o            | 4020   | 4658   | +15.9% | **0.86x**
LazyFilter.o              | 6536   | 7302   | +11.7% | **0.90x**
ReversedCollections.o     | 7090   | 7854   | +10.8% | **0.90x**
ArrayRemoveAll.o          | 7158   | 7904   | +10.4% | **0.91x**
CString.o                 | 5120   | 5629   | +9.9%  | **0.91x**
DictionaryGroup.o         | 10661  | 11680  | +9.6%  | **0.91x**
BucketSort.o              | 7373   | 7976   | +8.2%  | **0.92x**
QueueTest.o               | 11241  | 11993  | +6.7%  | **0.94x**
RemoveWhere.o             | 12027  | 12785  | +6.3%  | **0.94x**
StringEdits.o             | 8851   | 9316   | +5.3%  | **0.95x**
ArrayAppend.o             | 22864  | 23942  | +4.7%  | **0.95x**
WordCount.o               | 34407  | 35428  | +3.0%  | **0.97x**
Breadcrumbs.o             | 35631  | 36396  | +2.1%  | **0.98x**
StringComparison.o        | 30562  | 31090  | +1.7%  | **0.98x**
FloatingPointConversion.o | 34830  | 35385  | +1.6%  | **0.98x**
ObjectiveCBridgingStubs.o | 15351  | 15567  | +1.4%  | **0.99x**
SetTests.o                | 124525 | 125854 | +1.1%  | **0.99x**
&nbsp; | | | | 
**Improvement**           | **OLD** | **NEW** | **DELTA** | **RATIO**
IntegerParsing.o          | 61902  | 59772  | -3.4%  | **1.04x**
AngryPhonebook.o          | 8344   | 8128   | -2.6%  | **1.03x**
DriverUtils.o             | 132334 | 131006 | -1.0%  | **1.01x**

### Performance (x86_64): -Onone

**Regression**                          | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                                    | ---:    | ---:    | ---:   | ---:     
EqualStringSubstring                    | 27.353  | 35.407  | +29.4% | **0.77x**
EqualSubstringString                    | 28.303  | 35.6    | +25.8% | **0.80x (?)**
UTF8Decode_InitFromData                 | 139.1   | 174.556 | +25.5% | **0.80x (?)**
LessSubstringSubstringGenericComparable | 25.852  | 32.226  | +24.7% | **0.80x (?)**
LessSubstringSubstring                  | 27.333  | 34.032  | +24.5% | **0.80x (?)**
EqualSubstringSubstringGenericEquatable | 26.462  | 32.942  | +24.5% | **0.80x (?)**
UTF8Decode_InitFromBytes                | 143.667 | 177.667 | +23.7% | **0.81x (?)**
EqualSubstringSubstring                 | 27.893  | 34.364  | +23.2% | **0.81x (?)**
CSVParsing.Scalar                       | 508.25  | 606.0   | +19.2% | **0.84x (?)**
Breadcrumbs.MutatedUTF16ToIdx.Mixed     | 209.1   | 240.778 | +15.1% | **0.87x (?)**
Breadcrumbs.MutatedIdxToUTF16.Mixed     | 213.889 | 244.333 | +14.2% | **0.88x (?)**
ObjectiveCBridgeStringHash              | 62.261  | 69.391  | +11.5% | **0.90x (?)**
BridgeString.find.native.longNonASCII   | 383.0   | 424.75  | +10.9% | **0.90x (?)**
Set.isSubset.Seq.Int100                 | 3229.0  | 3499.0  | +8.4%  | **0.92x (?)**
&nbsp; | | | | 
**Improvement**                         | **OLD** | **NEW** | **DELTA** | **RATIO**
NSStringConversion.InlineBuffer.ASCII   | 6029.0  | 5095.0  | -15.5% | **1.18x (?)**
NSStringConversion.InlineBuffer.UTF8    | 3594.0  | 3172.0  | -11.7% | **1.13x (?)**
ArrayAppendAsciiSubstring               | 24354.0 | 22608.0 | -7.2%  | **1.08x (?)**

### Code size: -swiftlibs

**Improvement**                      | **OLD** | **NEW** | **DELTA** | **RATIO**
:---                                 | ---:  | ---:  | ---:   | ---:   
libswiftSwiftPrivateLibcExtras.dylib | 32768 | 16384 | -50.0% | **2.00x**


<details>
  <summary><strong>How to read the data</strong></summary>
The tables contain differences in performance which are larger than 8% and
differences in code size which are larger than 1%.

If you see any unexpected regressions, you should consider fixing the
regressions before you merge the PR.

**Noise**: Sometimes the performance results (not code size!) contain false
alarms. Unexpected regressions which are marked with '(?)' are probably noise.
If you see regressions which you cannot explain you can try to run the
benchmarks again. If regressions still show up, please consult with the
performance team (@eeckstein).
</details>

<details>
  <summary><strong>Hardware Overview</strong></summary>

      Model Name: Mac mini
      Model Identifier: Macmini8,1
      Processor Name: 6-Core Intel Core i7
      Processor Speed: 3.2 GHz
      Number of Processors: 1
      Total Number of Cores: 6
      L2 Cache (per Core): 256 KB
      L3 Cache: 12 MB
      Memory: 32 GB

</details>

DougGregor commented:
Oof, gonna need to teach the optimizers how to deal with typed throws.

DougGregor commented:
@swift-ci please benchmark

stephentyrone commented:
Do we understand where the large changes in the two MapReduce benchmarks come from?

eeckstein commented:
It's probably because function signature optimization is disabled for indirect throws (Though I didn't verify that assumption)

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please benchmark

DougGregor commented:
I've rebased this PR and am re-running all of the tests, now that SE-0413 has been accepted and typed throws has been enabled by default

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Debug info] Map the thrown error type into context when generating debug info

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
Update ABI test for typed-throws `map`

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
https://github.com/apple/swift-package-manager/pull/7249

@swift-ci please test source compatibility

DougGregor commented:
The swift-testing failure involves the C++ standard library; nothing to do with typed throws. Time to merge!

DougGregor commented:
> By any chance do you remember why specifically `map` was chosen for this migration? Do you remember why APIs like `filter` are still only `rethrows`? Do you remember any more examples that landed to show what this migration looks like and what steps look like to do that as backward compatible as possible? Thanks!

We're slowly going through the standard library to do the typed-throws updates. Here's a recent PR that did it in the concurrency library: https://github.com/swiftlang/swift/pull/85518
